### PR TITLE
[RFC] New protected TLV format proposal

### PR DESCRIPTION
**Pull request by utzig**
Updates the protected TLV specification to have their own TLV info struct, to ease hashing without having to skip parts of the image, while adding the ability to append new non-protected TLVs.

### CODE REVIEW

**Comment by d3zd3z** [comment]
So, my concern with not having the header describe the length of the protected header is that it then seems like it might be easier to just remove the protected header. Not specifying what the length of the trailer is makes us susceptible to a length-extension attack (well, maybe, but it makes me leery). A flag that indicates that a protected trailer must be present is a bit better. Having the length also makes it a little easier to find the unprotected trailer.

**Comment by utzig** [comment]
> So, my concern with not having the header describe the length of the protected header is that it then seems like it might be easier to just remove the protected header.

Yes, that is a good point; it would still require the image creator to release a correctly signed version with the same header information, and no protected trailers, and then later release an update with the protected trailers added for it to be possible for someone to just remove them, no?

Anyway I think it's fine to leave the protected length in the header (and I don't need to update the proposal!).

File: docs/design.md [hunk 1]
```diff
@@ -874,22 +881,15 @@ producing signed images, see: [signed_images](signed_images.md).
...
 
-If the TLV area contains dependency TLV entries, then these are required to be
+If the TLV area contains protected TLV entries, by beginning with a `struct
+image_tlv_info` with a magic value of `IMAGE_TLV_PROT_INFO_MAGIC` then the
+data of those TLVs with size given by `ih_protect_tlv_size` must also be
 integrity and authenticity protected. In this case the SHA256 has to be
-calculated over not just the image header and the image but also the TLV info
-header and the dependency TLVs.
+calculated over not just the image header and the image but also the TLV
+info header and the dependency TLVs.
```
[the remark this answers is not in the document]
Comment (utzig) [reply docs/design.md]: The text was update to clarify the issue.

File: docs/design.md [hunk 2]
```diff
@@ -898,9 +903,11 @@ A +---------------------+
...
   | TLV area            |
   | +-----------------+ |
-  | | TLV area header | | <- struct image_tlv_info
+  | | TLV area header | | <- struct image_tlv_info (optional)
   | +-----------------+ |
-  | | Dependency      | | <- Dependency entry (struct image_tlv)
+  | | Protected TLVs  | | <- Protected TLVs (struct image_tlv)
+  | +-----------------+ |
+  | | TLV area header | | <- struct image_tlv_info
 B | +-----------------+ |
```
Comment (tamasban) [inline docs/design.md]: Should not be B in line 909? The unprotected TLV header should not be included.

File: docs/design.md [hunk 3]
```diff
@@ -898,9 +903,11 @@ A +---------------------+
   +---------------------+
   | TLV area            |
   | +-----------------+ |
-  | | TLV area header | | <- struct image_tlv_info
+  | | TLV area header | | <- struct image_tlv_info (optional)
```
Comment (tamasban) [inline docs/design.md]: might put also the protected TLV area header magic here

File: docs/design.md [hunk 4]
```diff
@@ -898,9 +903,11 @@ A +---------------------+
...
   | +-----------------+ |
-  | | Dependency      | | <- Dependency entry (struct image_tlv)
+  | | Protected TLVs  | | <- Protected TLVs (struct image_tlv)
+  | +-----------------+ |
+  | | TLV area header | | <- struct image_tlv_info
```
Comment (tamasban) [inline docs/design.md]: might put also the unprotected TLV area header magic here

File: docs/design.md (hunk 2 above)
Comment (utzig) [reply docs/design.md]: Yes, thanks for point this out! Fixed.

File: docs/design.md (hunk 3 above)
Comment (utzig) [reply docs/design.md]: Done.

File: docs/design.md (hunk 4 above)
Comment (utzig) [reply docs/design.md]: Done.

**Comment by utzig** [comment]
Thanks for the reviews, I added this commit to #551, and will add reviewers when implementation is done.

**Pull request closed without merging**